Add eclipse plugins

rycee opened the pull request:
This PR adds a function to create a `bin/eclipse` file for a given Eclipse derivation such that a given list of plugins are available to that Eclipse. A number of initial plugins are provided. Example usage:

``` nix
{
  myEclipse = with eclipses; eclipseWithPlugins {
    eclipse = eclipse_sdk_45;
    plugins = with plugins; [
      anyedittools emacsplus
      findbugs color-theme
      checkstyle eclemma testng
    ];
  };
}
```

rycee committed:
eclipses: add function eclipseWithPlugins

This function allows installation of a given Eclipse with a given list
of Eclipse plugins.

rycee committed:
eclipse-anyedittools: init at 2.4.15.201504172030

This commit also introduces some infrastructure surrounding Eclipse
plugins.

rycee committed:
eclipse-emacsplus: init at 4.2.0

rycee committed:
eclipse-findbugs: init at 3.0.1.20150306-5afe4d1

rycee committed:
eclipse-color-theme: init at 1.0.0.201410260308

rycee committed:
eclipse-checkstyle: init at 6.5.0.201504121610

Also add helper function for building packages from downloaded Eclipse
update site.

rycee committed:
eclipse-eclemma: init at 2.3.2.201409141915

rycee committed:
eclipse-testng: init at 6.9.5.201506120235

bjornfor commented:
Cool! I've wanted this for a long time :-)

Why is it marked as WIP (work-in-progress)? What is missing?

edolstra commented:
@bjornfor Documentation (especially needed when packages cannot be installed in the "standard" way, i.e. via `nix-env -i`).

rycee committed:
eclipse-jdt: init at 4.5

rycee committed:
eclipse-scala: init at 4.1.0

rycee committed:
eclipse-platform: init at 4.5

rycee commented:
Yeah, this will need some documentation. Also I'm still not entirely certain whether the approach taken by the `eclipseWithPlugins` function is the best. I think it is pretty convenient now that I have used it for a few days and I especially like that the `eclipse` executable is the only thing exposed in my Nix environment. But perhaps it could be better?

One improvement that I would like to make for `eclipseWithPlugins` is to support plugins that provide their own executables. In particular I'm thinking about [Eclim](http://eclim.org/).

By the way, while the WIP is in progress I reserve the right to force push at will. I have, for example force pushed in a change that allows to set custom JVM arguments in the Eclipse ini.

rycee committed:
eclipses.plugins: add `buildEclipsePluginBase`

This function provides functionality common to all Eclipse plugin
builders. In particular, it sets a package name and flags the derivation
as an Eclipse plugin.

bjornfor commented:
How about exposing buildEclipsePlugin and buildEclipseUpdateSite to end-users? (Maybe under pkgs.eclipses.plugins.*?) That would make it easy to use out-of-tree plugins.

rycee commented:
@bjornfor, good idea. I will make that change.

rycee committed:
eclipses.plugins: put standard builders in set

This makes these builder functions available for use outside the file.
That is, plugins can be defined out-of-tree.

rycee committed:
eclipse-acejump: init at 1.0.0.201501181511

rycee committed:
eclipses: do not recurse into plugins.nix

It does not really make sense to install the plugin packages directly as
they are intended for use with `eclipseWithPlugins`. Therefore it is
best not to present them to users as such.

bjornfor commented:
Suggestion: make "meta" optional (for example, "meta ? {}") in buildEclipsePlugin and buildEclipseUpdateSite. This is mostly a convenience for out-of-tree plugins.

bjornfor commented:
buildEclipseUpdateSite takes a "version" parameter, but does not use it.

rycee committed:
eclipse.plugins: remove useless function arguments

Simplifies `buildEclipsePlugin` and `buildEclipseUpdateSite` functions
such that they require only absolutely necessary arguments. Also
add/expand comments slightly.

rycee committed:
scala-eclipse: 4.1.0 -> 4.1.1

rycee commented:
Again thanks for the good comments. I've added a commit to minimize the number of arguments these two functions require.

I've also started writing a bit of documentation in the nixpkgs manual, as a new subsection "Eclipse" in the "Package Notes" section. Though I haven't gotten very far on that yet.

bjornfor commented:
@rycee: Good. Btw, I think personally I'd keep the "src" parameter mandatory, because I don't think there is a use case for using the function _without_ "src" and it makes it easier to understand how to use the function.

bjornfor commented:
There is a plugin that does not work with this PR, it's called GNU ARM Eclipse (http://gnuarmeclipse.livius.net/blog/). A plugin zip file can be downloaded from sourceforge[1] -- this is the file I use. I'm not sure if this plugin is packaged in some weird way, or if simply this PR's way of handling plugins does not work for all.

Now, to the problem. When creating a new project with the GNU ARM Eclipse plugin[2], the project creation wizard fails at the end when trying to copy a template file:

```
Condition 7--> Process 1 (ilg.gnuarmeclipse.templates.core.ConditionalAddFiles)-->Error: Add Files failure: template source not found: file:/nix/store/6fir1p91y0nk6svpghyb6yzjyia2klvb-eclipse-cpp-4.4/eclipse/../../6d35mpm7a7kv1g00nz8hpi8533gaf9ya-eclipse-gnuarmeclipse-2.8.1-201504061754/eclipse/dropins/gnuarmeclipse-2.8.1-201504061754/plugins/ilg.gnuarmeclipse.templates.core_2.4.1.201504061754.jar!/templates/common//src/main/main-100-include.c
```

The .jar file in the error message exists, and within it there is also this /templates/common/src/main/main-100-include.c file that is also referred to (I'm guessing the double-slash in the error message is ignored). So as far as I can see, it should have been working. But it is not.

I compared the plugin directory created with buildEclipseUpdateSite and the official way (Eclipse -> Help -> Install New Software... -> Add... -> Archive...). With buildEclipseUpdateSite I get directories only in features/ and .jar files only in plugins/. But when installed the official way, the plugin (or eclipse) even has converted some (but not all) of the plugins/NAME.jar files into plugins/NAME/ directories. In particular for this plugin, it has extracted the \*.templates.core_\*.jar file that the error message refers to. Installing the plugin the official way works fine, the template file is found.

Hm, what to do? Does the plugin contain info about whether some .jar files should be extracted?

I think I remember seeing somewhere that eclipse can be invoked with some arguments to install a plugin, basically the programmatic way to do "Eclipse -> Help -> Install New Software... -> Add... -> Archive...". Or am I dreaming?

[1] http://cznic.dl.sourceforge.net/project/gnuarmeclipse/Current%20Releases/2.x/ilg.gnuarmeclipse.repository-2.8.1-201504061754.zip
[2] Create new "C Project", select any of the "STM32..." alternatives and press enter a couple of times (accept defaults) to complete the wizard.

bjornfor commented:
Playing around with info from this link: http://stackoverflow.com/questions/23946703/eclipse-commandline-to-install-plugin-from-remote-zip (also works with local zip).

Basically:

```
/nix/store/6fir1p91y0nk6svpghyb6yzjyia2klvb-eclipse-cpp-4.4/bin/eclipse \
  -application org.eclipse.equinox.p2.director -noSplash \
  -repository file:///home/bfo/nixpkgs/ilg.gnuarmeclipse.repository-2.8.1-201504061754/ \
  -installIU ilg.gnuarmeclipse.packs.core
```

Issues:
* This installs into ~/.eclipse/org.eclipse.platform_4.4.0/plugins/. Need to find a way to install to arbitrary directory. -help reveals plenty of options, hopefully it already exist.
* Each "IU" must be specified on the command line. Need to find a way to install "everything". Possible hack: use -list operation to get all IUs and use shell to transform it into a comma-separated list for -installIU.

It's interesting to note that the -list operation on GNU ARM Eclipse reveals that some "IUs" are "jar" and some are not:

```
$ /nix/store/6fir1p91y0nk6svpghyb6yzjyia2klvb-eclipse-cpp-4.4/bin/eclipse -application org.eclipse.equinox.p2.director -noSplash -repository file:///home/bfo/nixpkgs/ilg.gnuarmeclipse.repository-2.8.1-201504061754/ -list
OpenJDK 64-Bit Server VM warning: ignoring option MaxPermSize=256m; support was removed in 8.0
201504061754.ilg.gnuarmeclipse.category=1.0.0.7X7s7icNPLHh8Ecy9dreSyo_r
ilg.gnuarmeclipse.codered=1.1.1.201504061754
ilg.gnuarmeclipse.codered.feature.group=1.1.1.201504061754
ilg.gnuarmeclipse.codered.feature.jar=1.1.1.201504061754
ilg.gnuarmeclipse.core=2.8.1.201504061754
ilg.gnuarmeclipse.debug.core=1.2.1.201504061754
ilg.gnuarmeclipse.debug.gdbjtag=2.2.1.201504061754
ilg.gnuarmeclipse.debug.gdbjtag.jlink=3.2.1.201504061754
ilg.gnuarmeclipse.debug.gdbjtag.jlink.feature.group=3.2.1.201504061754
ilg.gnuarmeclipse.debug.gdbjtag.jlink.feature.jar=3.2.1.201504061754
ilg.gnuarmeclipse.debug.gdbjtag.openocd=3.2.1.201504061754
ilg.gnuarmeclipse.debug.gdbjtag.openocd.feature.group=3.2.1.201504061754
ilg.gnuarmeclipse.debug.gdbjtag.openocd.feature.jar=3.2.1.201504061754
ilg.gnuarmeclipse.debug.gdbjtag.qemu=1.1.1.201504061754
ilg.gnuarmeclipse.debug.gdbjtag.qemu.feature.group=1.1.1.201504061754
ilg.gnuarmeclipse.debug.gdbjtag.qemu.feature.jar=1.1.1.201504061754
ilg.gnuarmeclipse.debug.gdbjtag.restart=1.3.1.201504061754
ilg.gnuarmeclipse.debug.packs=1.1.1.201504061754
ilg.gnuarmeclipse.doc.user=1.1.1.201504061754
ilg.gnuarmeclipse.doc.user.feature.group=1.1.1.201504061754
ilg.gnuarmeclipse.doc.user.feature.jar=1.1.1.201504061754
ilg.gnuarmeclipse.managedbuild.cross=1.13.1.201504061754
ilg.gnuarmeclipse.managedbuild.cross.feature.group=1.13.1.201504061754
ilg.gnuarmeclipse.managedbuild.cross.feature.jar=1.13.1.201504061754
ilg.gnuarmeclipse.managedbuild.packs=1.2.2.201504061754
ilg.gnuarmeclipse.packs=1.2.3.201504061754
ilg.gnuarmeclipse.packs.core=1.1.2.201504061754
ilg.gnuarmeclipse.packs.data=1.1.3.201504061754
ilg.gnuarmeclipse.packs.feature.group=2.2.2.201504061754
ilg.gnuarmeclipse.packs.feature.jar=2.2.2.201504061754
ilg.gnuarmeclipse.packs.ui=1.2.1.201504061754
ilg.gnuarmeclipse.templates.core=2.4.1.201504061754
ilg.gnuarmeclipse.templates.cortexm=1.2.5.201504061754
ilg.gnuarmeclipse.templates.cortexm.feature.group=1.2.5.201504061754
ilg.gnuarmeclipse.templates.cortexm.feature.jar=1.2.5.201504061754
ilg.gnuarmeclipse.templates.freescale=2.2.5.201504061754
ilg.gnuarmeclipse.templates.freescale.feature.group=2.2.5.201504061754
ilg.gnuarmeclipse.templates.freescale.feature.jar=2.2.5.201504061754
ilg.gnuarmeclipse.templates.freescale.pe=1.2.1.201504061754
ilg.gnuarmeclipse.templates.stm=2.2.5.201504061754
ilg.gnuarmeclipse.templates.stm.feature.group=2.2.5.201504061754
ilg.gnuarmeclipse.templates.stm.feature.jar=2.2.5.201504061754
Operation completed in 57 ms.
```

In particular, the troublesome ilg.gnuarmeclipse.templates.core from my previous message, is not listed as a a "jar". Playing around with installing these IUs, I found that Installing a "jar" thing leaves plugins/NAME.jar, installing a non-jar thing leaves plugins/NAME/ (directory).

Inspecting the jar/non-jar state of IUs might be a way forward, although it feels hackish. If we can make eclipse install the plugin itself, into an arbitrary directory, I think that'd be the best.

bjornfor commented:
No... wait. It seems the jar/non-jar output from -list does not correspond 100 % with whether the final output is plugin/NAME.jar or plugin/NAME/. I'll do some more testing.

rycee committed:
eclipse.plugins: unpack plugins when required

Some plugin bundles must be unpacked when used in Eclipse. With this
change the plugin manifest is checked for the setting indicating that
unpacking should happen.

rycee commented:
@bjornfor, I had hoped to avoid worrying about unpacking bundles but I guess it was inevitable :-) In any case, I've added a commit now that should enable unpacking for those plugin JARs that indicate (through their manifest file) that they want to be unpacked. Please give it a try with gnuarm, I did a basic try myself and I was at least able to create a project using the wizard.

I think installing through Eclipse itself would be nice but I'm a little worried that it would make the installation have a runtime dependency on the Eclipse that was used to install it. This might be an unfounded worry, though, and even if true probably wouldn't be a big problem.

bjornfor commented:
@rycee: Awesome! Thank you!

rycee committed:
nixpkgs doc: add package notes for Eclipse

rycee commented:
Haven't really had time to work on this the last couple of days but I've now finished up a bit of documentation. Please have a look at see if it makes sense. If it is OK I think I'm OK with moving this out of WIP and hopefully have it merged if no further issues are found.

bjornfor commented:
Looks good to me.

rycee commented:
Thanks! I didn't actually know that the wildcard is optional. Removed it from the documentation.

bjornfor commented on doc/package-notes.xml:
```diff
@@ -219,5 +219,80 @@ you should modify
+
+    Note that if there is a need to install plugins that are not
+    available in Nixpkgs then it is still possible to install them
+    using the regular Eclipse methods.
```
I feel we should also mention how to declaratively install out-of-tree plugins. I suggest the paragraph above is changed into something along these lines:

Plugins that are not in nixpkgs can be installed either from within Eclipse (the traditional/imperative way) or declaratively by using one of the ```buildEclipsePlugin``` or ```buildEclipseUpdateSite``` functions. Which function to use depends on whether the plugin is distributed as separate JAR files or as a complete update site archive.

Below is an example that builds an Eclipse environment including the color-theme plugin from nixpkgs, and an out-of-tree plugin:

```
packageOverrides = pkgs: {

  myEclipse = with pkgs.eclipses; eclipseWithPlugins {
    eclipse = eclipse_cpp_45;
    plugins = [
      plugins.color-theme
      (plugins.buildEclipseUpdateSite rec {
        name = "my-plugin-1.0";
        src = fetchzip {
          stripRoot = false;
          url = "http://..."
          sha256 = "123abc...";
        };
      })
    ];
  };

};
```

Thoughts?

rycee replied:
Yeah, that is a good point. I added some text about this and an example with both `buildEclipseUpdateSite` and `buildEclipsePlugin`.

bjornfor commented on doc/package-notes.xml:
```diff
@@ -219,5 +219,80 @@ you should modify
+  </para>
+
+  <para>
+    This installation method provides an Eclipse installation that
```
I was confused at first because I thought the "This" in this paragraph referred to the text that _followed_. Maybe I'm wrong, but I think it is better if the sentence "This installation method provides an Eclipse installation that [...]" is moved to the previous paragraph (where the "This" actually refers to).

rycee replied:
Agreed, I've fixed this.

bjornfor commented:
@rycee: Very nice! (Much better than what I wrote.)

Do you want to do some commit squashing before I merge?

rycee commented:
@bjornfor: Thanks for the praise. I think the commits generally make sense on their own without needing to be squashed further. I did squash one commit with only a few fixed comments. But unless there is some specific concern I think I'm happy with them as-is.

bjornfor commented:
Pushed to master (commit c71d44c724be7ca87c5e426bab350b7261678430 with parents). Thank!

joelmo commented:
@rycee did you try to package Eclim? I want it but can't find it in nixpkgs, then I found this post.

rycee commented:
@joelmo, unfortunately I never made a serious effort to package Eclim. It is not as straight forward as most other Eclipse plugins and I just make due with the Emacs+ Eclipse plugin.
